[Inductor] Defer reduction split after fusion

y-sq opened the pull request:
**Context**
We mainly want to address the two cases that are commonly used in float8 training:
1. reduction with a tiled pointwise, for example:
```
def test_1(input_x):
    y = input_x.abs().max()

    z = input_x / 10.0
    z_t = z.t().contiguous().t()  # `z` and `z_t` will be fused into a tiled pointwise

    return y, z, z_t
```
This fusion of `y` and `z_z_t` failed due to "invalid tiling" (https://github.com/pytorch/pytorch/issues/128063)

2. reduction with a "partial reduction", for example:
```
def test_2(x):
    y = x.abs().max(dim=-1)
    z = x.abs().max()   # we want the first-level reduction of `z` can be fused with `y`.
    return y[0], z
```
This fusion doesn't happen also because `y` and first-level reduction of `z` has a different tiling (https://github.com/pytorch/pytorch/issues/136267)

To fix the two cases, we attempted to defer the split of reduction after fusion, so that the reductio split can take its neighbor op into consideration to allow the fusions to happen.

**Summary of the changes**:
* Added a config option `defer_reduction_split`. When it's enabled, if `num_splits` gets a `>1` result, return `ReductionHint.DEFERRED_SPLIT, 1` instead.
* In scheduler, when fusing nodes, if a node is a `ReductionHint.DEFERRED_SPLIT`, ignore it for now. After all non-DEFERRED_SPLIT nodes are fused, split DEFERRED_SPLIT Reduction nodes.
  * I thought about a better place to put `split_reduction_nodes`. However, there are cases that the tiled operations will only occur after fusion. So I chose to run `fuse_nodes` first, and then used the fused nodes to split the reduction nodes.
For example,
```
# op0. the reduction, read: arg0_1
y = x.abs().max()
# op1. read: arg0_1; write: buf0; no tiling
> z = x / 10.0   
# op2. read: transposed buf0              
z_t = z.t().contiguous().t()  

# Before fusion of op1 and op2:
Only op1 has shared data with op0, however op1 has no tiling
# After fusion of op1 and op2:
fused op0_1: read arg0_1, and has a tiling that the reduction (op0) can try to match
```
* For a  DEFERRED_SPLIT Reduction node, find all nodes that has shared data with it. (if `loop_reordering_after_fusion` is enabled, `loop_reordering` will also be considered.) 
Use the other nodes' tiling as the hints to split the reduction. If "hints" are too different from the original num_split, they'll be ignored.
*  Create new Scheduler nodes from the split reduction nodes. Remove the DEFERRED_SPLIT Reduction node and add the new Reduction nodes in Scheduler, and updating all related scheduler fields. (Hope I didn't miss any necessary fields in the scheduler...)
*  If there are new nodes added, re-run `fuse_nodes`. The second `fuse_nodes` should finish quickly as only the newly added reduction nodes may be fused into other nodes.

**Test Plan**:
Some test cases in my testing scripts in fbcode (in the diff)
```
TORCHINDUCTOR_PROFILE_OUTPUT=/tmp/profile.txt TORCHINDUCTOR_PROFILE=1 TORCHINDUCTOR_PROFILE_WITH_DO_BENCH_USING_PROFILING=1  TORCHINDUCTOR_DEFER_REDUCTION_SPLIT=1 TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=1 TORCH_LOGS="fusion" buck run mode/opt scripts/shuqiyang/test_inductor:test_reduction 2>&1 | tee ~/test_reduction_log_4.txt
```
* Includes both the delayed scaling and luca's rowwise scaling. Next step is to test it in actual float8 training use case.
------
Unit test:
Defer_reduction_split test:
```
buck2 test 'fbcode//mode/opt' caffe2/test/inductor:defer_reduction_split
```

Differential Revision: D66157846




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

y-sq commented on torch/_inductor/scheduler.py:
```diff
@@ -3318,7 +3464,7 @@ def score_fusion_memory(
 
         # optimization: iter over smaller set
-        if max(node1_dep_len, node2_dep_len) * 4 > min(node1_dep_len, node2_dep_len):
+        if min(node1_dep_len, node2_dep_len) * 4 < max(node1_dep_len, node2_dep_len):
```
Unrelated to this pr, but I guess it was a typo?

y-sq commented on torch/_inductor/scheduler.py:
```diff
@@ -3072,14 +3209,23 @@ def can_fuse(self, node1: BaseSchedulerNode, node2: BaseSchedulerNode) -> bool:
+            shared_data_score,
+        )
+
+        if shared_data_score < config.score_fusion_memory_threshold:
```
Unrelated to this pr, it's to fix this test error, https://www.internalfb.com/intern/test/844425111960859?ref_report_id=0

y-sq commented on torch/_inductor/scheduler.py:
```diff
@@ -1902,6 +1924,114 @@ def create_scheduler_node(self, node: ir.Operation) -> BaseSchedulerNode:
+                self.name_to_buf.update({
+                    buf.get_name() : buf for buf in middle_node.get_outputs()
+                })
+                self.name_to_fused_node[middle_node.get_name()] = middle_node
```
Hope I didn't miss to update other necessary fields in scheduler...

y-sq commented:
@pytorchbot label "topic: not user facing"

vkuzo commented:
Nice! I tried this on a torchao test case and get the following error:

```
> cd torchao
> TORCHINDUCTOR_DEFER_REDUCTION_SPLIT=1 TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=1 python benchmarks/float8/profile_linear_float8.py ~/local/tmp/20241120_test --dtype_filter float8 --scaling_type_input delayed --scaling_type_weight delayed --scaling_type_grad_output delayed
...
  File "/data/users/vasiliy/pytorch/torch/_inductor/scheduler.py", line 1778, in __init__
    self._init(nodes)
  File "/data/users/vasiliy/pytorch/torch/_inductor/scheduler.py", line 1851, in _init
    has_new_split_nodes = self.split_reduction_node()
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/users/vasiliy/pytorch/torch/_inductor/scheduler.py", line 2031, in split_reduction_node
    self.compute_ancestors()
  File "/data/users/vasiliy/pytorch/torch/_inductor/scheduler.py", line 2406, in compute_ancestors
    ancestors |= name_to_ancestors[dep_node_name]
                 ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
KeyError: 'op8'
```

y-sq commented:
@vkuzo  thanks for sharing the error. Let me look into it.

y-sq commented on torch/_inductor/scheduler.py:
```diff
@@ -3087,14 +3261,26 @@ def can_fuse(self, node1: BaseSchedulerNode, node2: BaseSchedulerNode) -> bool:
         del device2
 
         shared_data_score = self.score_fusion_memory(node1, node2)
-        if shared_data_score == 0:
```
Unrelated to this pr, it's to fix this test error, https://www.internalfb.com/intern/test/844425111960859?ref_report_id=0

y-sq replied:
The main change here is to modify the condition of calling loop_reordering from `shared_data_score == 0` to `shared_data_score < config.score_fusion_memory_threshold`. 

In the failed test (https://github.com/pytorch/pytorch/blob/main/test/inductor/test_loop_ordering.py#L425), the two pointwise nodes originally have a shared read data (the scaling factor, which is just a scalar). The `shared_data_score` is not `0` but smaller than `config.loop_ordering_after_fusion`. 

Before the change:
`shared_data_score > 0` -> won't `loop_reorder` -> can't fused because of `shared_data_score < config.score_fusion_memory_threshold`
After the change:
`shared_data_score > 0` -> `loop_reorder` (shared_data_score < config.score_fusion_memory_threshold`) -> get a larger `shared_data_score` -> fused

y-sq committed:
[Inductor] Defer reduction split after fusion (#141082)

Summary:

* Added a config option `defer_reduction_split`. When it's enabled, if `num_splits` gets a `>1` result, return `ReductionHint.DEFERRED_SPLIT, 1` instead.
* In scheduler, when fusing nodes, if a node is a `ReductionHint.DEFERRED_SPLIT`, ignore it for now. After all non-DEFERRED_SPLIT nodes are fused, split DEFERRED_SPLIT Reduction nodes.
  * I thought about a better place to put `split_reduction_nodes`. However, there are cases that the tiled operations will only occur after fusion. So I chose to run `fuse_nodes` first, and then used the fused nodes to split the reduction nodes.
For example,
```
# op0. the reduction, read: arg0_1
y = x.abs().max()
# op1. read: arg0_1; write: buf0; no tiling
> z = x / 10.0   
# op2. read: transposed buf0              
z_t = z.t().contiguous().t()  

# Before fusion of op1 and op2:
Only op1 has shared data with op0, however op1 has no tiling
# After fusion of op1 and op2:
fused op0_1: read arg0_1, and has a tiling that the reduction (op0) can try to match
```
* For a  DEFERRED_SPLIT Reduction node, find all nodes that has shared data with it. (if `loop_reordering_after_fusion` is enabled, `loop_reordering` will also be considered.) 
Use the other nodes' tiling as the hints to split the reduction. If "hints" are too different from the original num_split, they'll be ignored.
*  Create new Scheduler nodes from the split reduction nodes. Remove the DEFERRED_SPLIT Reduction node and add the new Reduction nodes in Scheduler, and updating all related scheduler fields. (Hope I didn't miss any necessary fields in the scheduler...)
*  If there are new nodes added, re-run `fuse_nodes`. The second `fuse_nodes` should finish quickly as only the newly added reduction nodes may be fused into other nodes.

Test Plan:
ome test cases in my testing scripts
```
TORCHINDUCTOR_PROFILE_OUTPUT=/tmp/profile.txt TORCHINDUCTOR_PROFILE=1 TORCHINDUCTOR_PROFILE_WITH_DO_BENCH_USING_PROFILING=1  TORCHINDUCTOR_DEFER_REDUCTION_SPLIT=1 TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=1 TORCH_LOGS="fusion" buck run mode/opt scripts/shuqiyang/test_inductor:test_reduction 2>&1 | tee ~/test_reduction_log_4.txt
```
* Includes both the delayed scaling and luca's rowwise scaling. Next step is to test it in actual float8 training use case.
------
Unit test:
Defer_reduction_split test:
```
buck2 test 'fbcode//mode/opt' caffe2/test/inductor:defer_reduction_split
```

There was a test failure in loop_reordering, https://fburl.com/test/yo50vt6x. Related fusion logs: P1682598849. Also fixed it in this diff.
```
buck2 test 'fbcode//mode/opt' caffe2/test/inductor:loop_ordering
```
----
Test in Float8 Training
```
TORCHINDUCTOR_DEFER_REDUCTION_SPLIT=1 TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=1  buck run mode/opt scripts/shuqiyang/test_inductor:test_float8 --  ~/local/tmp/20241120_test --dtype_filter float8 --scaling_type_input delayed --scaling_type_weight delayed --scaling_type_grad_output delayed  2>&1 | tee ~/test_compile
```
Trace: https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/trace.64960216286.json.gz&bucket=acadia
Codegen kernels: P1683645463, P1683645752

Differential Revision: D66157846

vkuzo commented:
thanks @y-sq !  Without activation checkpointing, I now see what I think is the optimal kernel patterns generated for fwd+bwd of a Float8Linear, measured with

```
TORCHINDUCTOR_DEFER_REDUCTION_SPLIT=1 TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=1 python benchmarks/float8/profile_linear_float8.py ~/local/tmp/20241120_test --dtype_filter float8 --scaling_type_input delayed --scaling_type_weight delayed --scaling_type_grad_output delayed
```

If we add AC (requires https://github.com/pytorch/ao/pull/1354), then I still see a lot of missing fusions:

```
TORCHINDUCTOR_DEFER_REDUCTION_SPLIT=1 TORCHINDUCTOR_LOOP_ORDERING_AFTER_FUSION=1 python benchmarks/float8/profile_linear_float8.py ~/local/tmp/20241120_test --dtype_filter float8 --scaling_type_input delayed --scaling_type_weight delayed --scaling_type_grad_output delayed --enable_activation_checkpointing True
```

y-sq commented:
@vkuzo thanks for sharing the SAC script. I'll check the SAC case.
In SAC, is the following the ideal case for delayed scaling?

Avtivation:
- In fwd, amax + cast (without transpose); amax is save for bwd, and fp8 tensors are not
- In bwd, transposed_cast (without amax because it's saved from fwd)

Weights:
- In FSDP case, should be the same as activation - fp8 weights are not saved for bwd.

Gradients:
- Not affected by SAC

vkuzo commented:
yep that sounds right!

this is what I see now with SAC:

## forward

seems that max(abs(tensor)) is not fused with the cast to float8

![Screenshot 2024-11-26 at 2 31 49 PM](https://github.com/user-attachments/assets/dcc5926d-7d6b-444b-8247-dd8e5d10466b)

## backward

looks good (I think)

![Screenshot 2024-11-26 at 2 32 17 PM](https://github.com/user-attachments/assets/69d9c740-4fbc-44b3-802e-6712ce943072)

eellison commented on torch/_inductor/ir.py:
```diff
@@ -1291,6 +1334,8 @@ def create(
         reduction_type: str,
         reduction_hint: ReductionHint = ReductionHint.DEFAULT,
         input_node: Optional[IRNode] = None,
+        ignore_defer_reduction_split: bool = False,
```
I guess we are not yet supporting Welford, that's fine.

eellison commented on torch/_inductor/runtime/hints.py:
```diff
@@ -20,7 +20,8 @@ class ReductionHint(Enum):
     OUTER = 1
     OUTER_TINY = 2
-    DEFAULT = 3
+    DEFERRED_SPLIT = 3
```
It's a little weird for this to show up at runtime, since it's a compile time only construct.

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -1845,6 +1865,13 @@ def _init(self, nodes: List[ir.Operation]) -> None:
         if config._pre_fusion_custom_pass is not None:
             self.nodes = config._pre_fusion_custom_pass(self.nodes)
         self.nodes = self.fuse_nodes(self.nodes)
```
Hmm, I understand we want to do some amount of fusion prior to fusion of split nodes, however, the current fusion occurs so that we greedily fuse most profitable fusions first (saved global memory). And we are currently willing to do it for 10 iterations or until we did not do any fusions in the current round.

I'm not sure what the best approach is here but I don't think we want to do 20 total rounds. Potentially some smaller number of possible fusions, first, then split reductions, then fuse. And in first round we could consider just attempting fusing pointwise nodes.

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -2377,34 +2548,34 @@ def benchmark_fused_nodes(
+    def replace_operation_buffer(
+        self, orig_node: ir.OperationBuffer, new_node: ir.OperationBuffer
+    ) -> None:
+        replaced_buf_name = new_node.get_name()
```
nice, glad you found this !

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -1916,6 +1943,145 @@ def create_scheduler_node(self, node: ir.Operation) -> BaseSchedulerNode:
         else:
             raise NotImplementedError(node)
 
+    def split_reduction_node(self) -> bool:
```
Still need to review this function

eellison reviewed (commented):
Looks good ! a few comments. would you do a perf run on the dashboard ? see https://pytorch.org/docs/stable/torch.compiler_performance_dashboard.html

i still need to review one of the functions you added.

cc @jansel, @Chillee for further thoughts

jansel reviewed (changes requested):
Can you talk a bit more about the motivation for this?  Longer term I want to remove/disable split reductions in factor of cooperative reductions (which can fuse more things and works better with dynamic shapes).

Setting `triton.cooperative_reductions=True` might solve the same problem as this.

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -1916,6 +1943,145 @@ def create_scheduler_node(self, node: ir.Operation) -> BaseSchedulerNode:
+
+        def split_reduction_single_node(idx, snode, all_nodes) -> None:
+            split_hints = []
+            for other_node in all_nodes:
```
We shouldn't need to iterate over all_nodes. see, [here](https://github.com/pytorch/pytorch/blob/fd553b9817cb145e3cfaa52a08a831b9edf2fec7/torch/_inductor/scheduler.py#L2736-L2743) where we first group by buffer accesses.

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -1916,6 +1943,145 @@ def create_scheduler_node(self, node: ir.Operation) -> BaseSchedulerNode:
+                    tiling = SIMDScheduling.select_tiling(
+                        other_node.get_nodes(), numel1, rnumel1
+                    )
+                    split_hints.append(tiling[0])
```
why only `tiling[0]` here ?

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -1916,6 +1943,145 @@ def create_scheduler_node(self, node: ir.Operation) -> BaseSchedulerNode:
+            self.name_to_node[snode.get_name()] = new_scheduler_node
+            self.name_to_fused_node[snode.get_name()] = new_scheduler_node
+
+            # It's possible that there is no additional middle-layer nodes
```
meaning, we didn't split the reduction ? Can we check if there would be a split prior to creating a new reduction ?

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -1916,6 +1943,145 @@ def create_scheduler_node(self, node: ir.Operation) -> BaseSchedulerNode:
+                assert isinstance(middle_buffer, ir.ComputedBuffer)  # linter
+                middle_buffer.layout = middle_buffer.layout.as_fixed()
+                middle_node = self.create_scheduler_node(middle_buffer)
+                self.nodes.append(middle_node)
```
why we are appending this in a loop?

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -1916,6 +1943,145 @@ def create_scheduler_node(self, node: ir.Operation) -> BaseSchedulerNode:
+                )
+                self.name_to_fused_node[middle_node.get_name()] = middle_node
+
+            # The node whose users are the orig_node (un-split reduction) should be updated to be the first middle reduction node
+            for node in self.nodes:
+                for buf in node.get_outputs():
+                    for user in buf.users:
+                        if user.node.get_name() == snode.get_name():
+                            user.node = self.nodes[orig_num_scheduler_node]
+            # Update the dependencies (buf.usesrs) of the new reduction nodes
+            self.compute_dependencies(
+                self.nodes[orig_num_scheduler_node:] + [new_scheduler_node]
```
In each iteration here we're updating  nodes/dependencies. this might be slow. Maybe it would make sense to collect all the buffers we're going to split first, then split, then update once ?

eellison reviewed (commented):
@jansel the motivation https://github.com/pytorch/pytorch/issues/128063. i think for fp8 they were hoping to have a working solution in the nearer term. my understanding with  cooperative reductions is it wouldn't necessarily be strictly more performant and also would require upstream triton changes for cooperative launches

y-sq commented on torch/_inductor/ir.py:
```diff
@@ -1130,6 +1131,25 @@ def outer_reduction_splits(reduction_numel_hint, numel_hint):  # type: ignore[no
+            if not split_hints:
+                return split_num
+
+            min_split_size = num_threads * min_elements_per_thread // 8  # 1024
+            max_split_size = num_threads * max_elements_per_thread * 2  # 262144
```
`If there is already a partial reduction occurring, as with https://github.com/pytorch/pytorch/issues/136267, I think it always makes sense to do the same split.`
Yes, this makes sense. I think I should have two types of split_hints..? - If it comes from a partial reduction, always follow it; if it comes from other tiled ops, only follow it if the split is not two different from the optimal one?

y-sq commented on torch/_inductor/ir.py:
```diff
@@ -1130,6 +1131,25 @@ def outer_reduction_splits(reduction_numel_hint, numel_hint):  # type: ignore[no
+            max_split_size = num_threads * max_elements_per_thread * 2  # 262144
+
+            new_split_num = None
+            best_diff = 16
```
yes, I'll change to `ratio`. I was thinking to define a threshold (the current ratio 16 might not be an ideal one), if the new split is too different from the original split, we might want to ignore it. For example, the case 32x16777216 might be a bad one to split the reduction as 32, 1677721... ?

y-sq commented:
> How does this interact with the cooperative reductions from @jansel ?

@Chillee , currently this change should have no interaction with the cooperative reductions.
The option "defer_reduction_split" only takes effects when the reduction should be split. When `cooperative_reductions` is enabled (https://github.com/pytorch/pytorch/blob/main/torch/_inductor/config.py#L484), Reduction.num_splits would always have `should_split == False`, https://github.com/pytorch/pytorch/blob/main/torch/_inductor/ir.py#L1049 (`V.graph.has_feature(device, BackendFeature.REDUCE_TO_SINGLE_ELEMENT)` is `True` with `cooperative_reductions`, https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/triton.py#L3457)

but I'll run some local tests to see if `cooperative_reductions` itself is enough to handle the fp8 use case.

y-sq commented:
Hi @jansel and @eellison , thanks for your comments about the cooperative reductions. I did some tests to see if it can directly help the fp8 cases.

For context about our needs in float8 training, we have the two cases that we want to handle:
1. reduction with a tiled pointwise, for example:
```
def test_1(input_x):
    y = input_x.abs().max()

    z = input_x / 10.0
    z_t = z.t().contiguous().t()  # `z` and `z_t` will be fused into a tiled pointwise

    return y, z, z_t

test = torch.compile(test)
x = torch.randn(3072, 4096, device="cuda") / 10.0
y, z, z_t = test(x)
```
This fusion of `y` and `z_z_t` failed due to "invalid tiling" (https://github.com/pytorch/pytorch/issues/128063)

2. reduction with a "partial reduction", for example:
```
def test_2(x):
    y = x.abs().max(dim=-1)
    z = x.abs().max()   # we want the first-level reduction of `z` can be fused with `y`.
    return y[0], z

test = torch.compile(test)
x = torch.randn(3072, 4096, device="cuda")
z = test(x)
```
This fusion doesn't happen also because `y` and first-level reduction of `z` has a different tiling (https://github.com/pytorch/pytorch/issues/136267)



----------

I did some tests with `TORCHINDUCTOR_COOPERATIVE_REDUCTIONS=1`:

**Performance of a single reduction**
```
def test(x):
        z = x.abs().max()
        return z

test = torch.compile(test)
x = torch.randn(3072, 4096, device="cuda")
z = test(x)
```
With `TORCHINDUCTOR_COOPERATIVE_REDUCTIONS=1`, I got:
```
TRITON KERNELS BANDWIDTH INFO (/tmp/torchinductor_shuqiyang/tmpzov4m_tx/7u/c7ulsfjspmgzwufvmj6qik5zc2fedgdveqvzn7hxvv7sohhu3utl.py)
0.038ms    	0.050 GB 	 1329.40GB/s 	 100.00% 	 triton_unk_fused_abs_max_0
SUMMARY (/tmp/torchinductor_shuqiyang/tmpzov4m_tx/7u/c7ulsfjspmgzwufvmj6qik5zc2fedgdveqvzn7hxvv7sohhu3utl.py)
0.04ms   	 0.05 GB	 1329.40GB/s
```
Without `TORCHINDUCTOR_COOPERATIVE_REDUCTIONS`, using the split-reduction, I got:
```
TRITON KERNELS BANDWIDTH INFO (/tmp/torchinductor_shuqiyang/tmpyhnx8ovm/rk/crkiazuwvplph2xsey4jrrk454uw2imrohmlqkgoeswrqcncacqi.py)
0.029ms    	0.050 GB 	 1727.88GB/s 	 92.89% 	 triton_red_fused_abs_max_0
0.002ms    	0.000 GB 	    0.92GB/s 	 7.11% 	 triton_per_fused_abs_max_1
SUMMARY (/tmp/torchinductor_shuqiyang/tmpyhnx8ovm/rk/crkiazuwvplph2xsey4jrrk454uw2imrohmlqkgoeswrqcncacqi.py)
0.03ms   	 0.05 GB	 1605.07GB/s
```

It seems that the current performance of is worse than split_reduction. Does it need "the triton upstream update" or "a flag to cuda kernel launches" to fix the performance issue?

**The two fp8 cases**
In the first case, they still can't be fused:
```
scheduler.py:2491] [0/0] [__fusion] ===== attempting fusion (2/10): 2 nodes =====
scheduler.py:2781] [0/0] [__fusion] fuse_nodes_once, candidates:
scheduler.py:2783] [0/0] [__fusion]   SchedulerNode(name='op0'), Reduction(['[3072, 4096]', 'max', 'origins=OrderedSet([abs_1, max_1])'])
scheduler.py:2783] [0/0] [__fusion]   FusedSchedulerNode(nodes=op1_op2), snodes: ["SchedulerNode(name='op1'), Pointwise(['[3072, 4096]', 'origins=OrderedSet([div])'])", "SchedulerNode(name='op2'), Pointwise(['[4096, 3072]', 'origins=OrderedSet([clone])'])"]
scheduler.py:747] [0/0] [__fusion] cannot fuse op1_op2 with op0: invalid tiling for reduction
scheduler.py:2909] [0/0] [__fusion] found 0 possible fusions
scheduler.py:2498] [0/0] [__fusion] completed fusion round (2/10): fused 2 nodes into 2 nodes
scheduler.py:2498] [0/0] [__fusion] 
scheduler.py:2505] [0/0] [__fusion] ===== fusion complete (2 iterations) =====
```

In the second case, also can't be fused:
```
scheduler.py:2491] [0/0] [__fusion] ===== attempting fusion (1/10): 2 nodes =====
scheduler.py:2781] [0/0] [__fusion] fuse_nodes_once, candidates:
scheduler.py:2783] [0/0] [__fusion]   SchedulerNode(name='op0'), Reduction(['[4096]', 'max', 'origins=OrderedSet([abs_1, max_1])'])
scheduler.py:2783] [0/0] [__fusion]   SchedulerNode(name='op2'), Reduction(['[3072, 4096]', 'max', 'origins=OrderedSet([max_2, abs_2])'])
scheduler.py:747] [0/0] [__fusion] cannot fuse op0 with op2: numel/rnumel mismatch (reduce) (3072, 1), (4096, 12582912)
scheduler.py:2909] [0/0] [__fusion] found 0 possible fusions
scheduler.py:2498] [0/0] [__fusion] completed fusion round (1/10): fused 2 nodes into 2 nodes
scheduler.py:2498] [0/0] [__fusion] 
scheduler.py:2505] [0/0] [__fusion] ===== fusion complete (1 iterations) =====
```

jansel commented:
The other thing you could try is increasing/decreasing:
https://github.com/pytorch/pytorch/blob/ecbb8a880028e59920b013f64805a883ed2958f5/torch/_inductor/runtime/triton_heuristics.py#L1766
and:
https://github.com/pytorch/pytorch/blob/ecbb8a880028e59920b013f64805a883ed2958f5/torch/_inductor/runtime/hints.py#L16

That heuristic is the same ideas as the split factor in split reductions, and currently it doesn't have a well tuned heuristic.  I'd expect if you make RSPLIT match the split factor the performance will be the closer.

y-sq commented:
I tried different TRITON_MAX_RSPLIT. With a proper RSPLIT value, the performance is very close to split_reduction.

TRITON_MAX_RSPLIT = 256,
```
TRITON KERNELS BANDWIDTH INFO (/tmp/torchinductor_shuqiyang/tmpf4fc6aqq/fi/cfikuh67balrjcaj27uqvppm44ot7hk6vq4llisvdfwv2ke3ycs5.py)
0.032ms    	0.050 GB 	 1590.97GB/s 	 100.00% 	 triton_unk_fused_abs_max_0
SUMMARY (/tmp/torchinductor_shuqiyang/tmpf4fc6aqq/fi/cfikuh67balrjcaj27uqvppm44ot7hk6vq4llisvdfwv2ke3ycs5.py)
0.03ms   	 0.05 GB	 1590.97GB/s
```

TRITON_MAX_RSPLIT = 512,
```
TRITON KERNELS BANDWIDTH INFO (/tmp/torchinductor_shuqiyang/tmp3qkw7bjj/6a/c6azb7rdb5pfwclyamp3pqztdqqrxccpjsuwgheanfnws6z5ih72.py)
0.031ms    	0.050 GB 	 1609.81GB/s 	 100.00% 	 triton_unk_fused_abs_max_0
SUMMARY (/tmp/torchinductor_shuqiyang/tmp3qkw7bjj/6a/c6azb7rdb5pfwclyamp3pqztdqqrxccpjsuwgheanfnws6z5ih72.py)
0.03ms   	 0.05 GB	 1609.81GB/s
```

However, besides the performance issue, we still need additional efforts to make the fusion work for COOPERATIVE_REDUCTIONS..?

eellison commented:
```
def test_1(input_x):
    y = input_x.abs().max()

    z = input_x / 10.0
    z_t = z.t().contiguous().t()  # `z` and `z_t` will be fused into a tiled pointwise

    return y, z, z_t

test = torch.compile(test)
x = torch.randn(3072, 4096, device="cuda") / 10.0
y, z, z_t = test(x)
```

For this case, we create a tiled pointwise kernel and are unable to fuse it to the reduction because the pointwise node [tiling is not equal to](https://github.com/pytorch/pytorch/blob/dbd7b820dd5f35517aef4ffdb21017baf2b15cfd/torch/_inductor/codegen/simd.py#L1087-L1088)
```
(numel1, 1),
(numel2, rnumel2, 1),
```
 
That is the pointwise kernel is not loading over the tensor contiguously nor does the tiling match the output # of elems and the # of reduced of inputs for the reduction - (numel2, rnumel2).

We don't currently support tiling reductions but @blaine-rister has a pr out for it here: https://github.com/pytorch/pytorch/pull/137243. I think in the end state with cooperative and tiled reductions both turned on we will be able to fuse this in a single kernel.

That said, it might take a little while for this to be on by default. Since we are mostly targeting a specific workload maybe we could land am fx pass that is simpler and easier to maintain in the short term, either in tree or as custom pass.

We already have one pass that targets the shared partial reduction. For that case, if we're targeting the joint graph saved activations, we'd need to do it as an fx pass regardless.

For the two step reduction that enables fusion with the tiled pointwise op, would it be feasible to write this way in the user code temporarily ? potentially we could write a pass that looks for use uses with transposed tensors.

y-sq commented:
@eellison thanks for the quick response. yes, I'll then work on the fx pass solution, and likely land that to torchao repo...? 
(cc @vkuzo  the idea of the fx pass solution is also split the reduction to two-levels to allow Inductor for further fusion, so it will also handle the fusion of preceding ops.)
